lib/rules/prefer-destructuring.js

```diff
@@ -42,24 +60,32 @@ module.exports = {
     },
     create(context) {
 
-        let checkArrays = true;
-        let checkObjects = true;
+        let checkArraysInVariableDeclarator = true;
+        let checkObjectsInVariableDeclarator = true;
+        let checkArraysInAssignmentExpression = true;
+        let checkObjectsInAssignmentExpression = true;
         let enforceForRenamedProperties = false;
         const enabledTypes = context.options[0];
         const additionalOptions = context.options[1];
 
         if (enabledTypes) {
-            if (typeof enabledTypes.array !== "undefined") {
-                checkArrays = enabledTypes.array;
+            if (typeof enabledTypes.VariableDeclarator !== "undefined") {
+                const variableDeclarator = enabledTypes.VariableDeclarator;
+
+                checkArraysInVariableDeclarator = variableDeclarator.hasOwnProperty("array") ? variableDeclarator.array : checkArraysInVariableDeclarator;
+                checkObjectsInVariableDeclarator = variableDeclarator.hasOwnProperty("object") ? variableDeclarator.object : checkObjectsInVariableDeclarator;
             }
 
-            if (typeof enabledTypes.object !== "undefined") {
-                checkObjects = enabledTypes.object;
+            if (typeof enabledTypes.AssignmentExpression !== "undefined") {
+                const assignmentExpression = enabledTypes.AssignmentExpression;
+
+                checkArraysInAssignmentExpression = assignmentExpression.hasOwnProperty("array") ? assignmentExpression.array : checkArraysInAssignmentExpression;
+                checkObjectsInAssignmentExpression = assignmentExpression.hasOwnProperty("object") ? assignmentExpression.object : checkObjectsInAssignmentExpression;
             }
         }
```

Comment:
This config-checking logic is getting kind of complicated. I think having a variable for each config case won't scale very well as more options are added, especially if we restore support for the old top-level "array" option. Can we extract this into a function instead? For example:

```js
/**
 * @param {string} nodeType "AssignmentExpression" or "VariableDeclaration"
 * @param {string} destructuringType "array" or "object"
 * @returns {boolean} `true` if the destructuring type should be checked for the given node
 */
function shouldCheck(nodeType, destructuringType) {
  return context.options[0] &&
    context.options[0][nodeType] &&
    context.options[0][nodeType][destructuringType] === false;
}
```

Then rather than handling four different variables, the `performCheck` function, it could just use something like

```js
if (isArrayIndexAccess(node) && shouldCheck(reportNode.type, "array")) {
    // ...
}

// ...
```